From 33bbb706d4e2fbfb7e818325f09a1bfa9501d111 Mon Sep 17 00:00:00 2001
From: Icenowy Zheng <icenowy@aosc.io>
Date: Sat, 6 May 2017 17:13:22 +0800
Subject: [PATCH 50/80] drm: sun4i: add support for A64 TCONs

A64 SoC has a TCON0 with only channel0 and a TCON1 with only channel1.

Add support for them.

Signed-off-by: Icenowy Zheng <icenowy@aosc.io>
---
 drivers/gpu/drm/sun4i/sun4i_drv.c  |  4 +++-
 drivers/gpu/drm/sun4i/sun4i_tcon.c | 20 ++++++++++++++++++++
 2 files changed, 23 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/sun4i/sun4i_drv.c b/drivers/gpu/drm/sun4i/sun4i_drv.c
index 9578f43cce0a..d714cee9db49 100644
--- a/drivers/gpu/drm/sun4i/sun4i_drv.c
+++ b/drivers/gpu/drm/sun4i/sun4i_drv.c
@@ -189,7 +189,9 @@ static bool sun4i_drv_node_is_tcon(struct device_node *node)
 		of_device_is_compatible(node, "allwinner,sun6i-a31s-tcon") ||
 		of_device_is_compatible(node, "allwinner,sun8i-a33-tcon") ||
 		of_device_is_compatible(node, "allwinner,sun8i-h3-tcon") ||
-		of_device_is_compatible(node, "allwinner,sun8i-v3s-tcon");
+		of_device_is_compatible(node, "allwinner,sun8i-v3s-tcon") ||
+		of_device_is_compatible(node, "allwinner,sun50i-a64-tcon0") ||
+		of_device_is_compatible(node, "allwinner,sun50i-a64-tcon1");
 }
 
 static int compare_of(struct device *dev, void *data)
diff --git a/drivers/gpu/drm/sun4i/sun4i_tcon.c b/drivers/gpu/drm/sun4i/sun4i_tcon.c
index 270f09e381a5..c173d4ae00fe 100644
--- a/drivers/gpu/drm/sun4i/sun4i_tcon.c
+++ b/drivers/gpu/drm/sun4i/sun4i_tcon.c
@@ -654,6 +654,10 @@ static const struct sun4i_tcon_quirks sun8i_a33_quirks = {
 	.has_channel_0	= true,
 };
 
+static const struct sun4i_tcon_quirks sun8i_a83t_tcon0_quirks = {
+	.has_channel_0	= true,
+};
+
 static const struct sun4i_tcon_quirks sun8i_v3s_quirks = {
 	.has_channel_0	= true,
 };
@@ -662,6 +666,14 @@ static const struct sun4i_tcon_quirks sun8i_h3_quirks = {
 	.has_channel_1	= true,
 };
 
+static const struct sun4i_tcon_quirks sun50i_a64_tcon0_quirks = {
+	.has_channel_0	= true,
+};
+
+static const struct sun4i_tcon_quirks sun50i_a64_tcon1_quirks = {
+	.has_channel_1	= true,
+};
+
 static const struct of_device_id sun4i_tcon_of_table[] = {
 	{ .compatible = "allwinner,sun5i-a13-tcon", .data = &sun5i_a13_quirks },
 	{ .compatible = "allwinner,sun6i-a31-tcon", .data = &sun6i_a31_quirks },
@@ -669,6 +681,14 @@ static const struct of_device_id sun4i_tcon_of_table[] = {
 	{ .compatible = "allwinner,sun8i-a33-tcon", .data = &sun8i_a33_quirks },
 	{ .compatible = "allwinner,sun8i-h3-tcon", .data = &sun8i_h3_quirks },
 	{ .compatible = "allwinner,sun8i-v3s-tcon", .data = &sun8i_v3s_quirks },
+	{
+		.compatible = "allwinner,sun50i-a64-tcon0",
+		.data = &sun50i_a64_tcon0_quirks
+	},
+	{
+		.compatible = "allwinner,sun50i-a64-tcon1",
+		.data = &sun50i_a64_tcon1_quirks
+	},
 	{ }
 };
 MODULE_DEVICE_TABLE(of, sun4i_tcon_of_table);
-- 
2.13.6

